### DISCUSSION TITLE
Accessing the Store on the com.apple.SwiftUI.AsyncRenderer thread

### DISCUSSION THREAD

**Discussion by gerfarfel** [pointfreeco/swift-composable-architecture#2767] [Q&A] [answered]
In converting my code to use @ObservableState (TCA 1.7), I ran into the following runtime warning:
>         "Store.scope" was called on a non-main thread. …
> 
>         The "Store" class is not thread-safe, and so all interactions with an instance of \
>         "Store" (including all of its scopes and derived view stores) must be done on the main \
>         thread.

Setting a breakpoint in ComposableArchitecture.RootStore.threadCheck reveals that the store is being accessed from the **com.apple.SwiftUI.AsyncRenderer** thread.

This issue was referenced in the comments for discussion #2694 on this forum, but without a solution. (Sorry if I've missed other references to the problem.)

Here's some code that generates the issue. There are two Reducers and two Views. The "SquareBox" view shows a colored square and recognizes tap gestures. The "Gallery" view shows a grid of SquareBox views. Tapping on any of the boxes deletes that box from the gallery. That's it... The runtime warning is triggered when a box is deleted. 

Note that the tap gesture action is sent with animation. If the action is not animated, the runtime warning is not triggered.

```
import ComposableArchitecture
import SwiftUI

// MARK: - SquareBox

@Reducer
struct SquareBox: Reducer {
  @ObservableState
  struct State: Equatable, Identifiable {
    let id = UUID()
    var color: Color
  }

  enum Action {
    case tapGesture
  }

  var body: some ReducerOf<Self> {
    Reduce { state, action in
      switch action {
      case .tapGesture:
        return .none
      }
    }
  }
}

struct SquareBoxView: View {
  var store: StoreOf<SquareBox>

  var body: some View {
    store.color
      .aspectRatio(1, contentMode: .fit)
      .onTapGesture {
        store.send(.tapGesture, animation: .default) // no runtime warning if this animation is removed!
      }
  }
}

// MARK: - Gallery

@Reducer
struct Gallery: Reducer {
  @ObservableState
  struct State: Equatable {
    var squareBoxes: IdentifiedArrayOf<SquareBox.State> = [
      SquareBox.State(color: .red),
      SquareBox.State(color: .green),
      SquareBox.State(color: .blue),
      SquareBox.State(color: .yellow),
      SquareBox.State(color: .purple),
      SquareBox.State(color: .orange),
      SquareBox.State(color: .cyan),
      SquareBox.State(color: .brown),
      SquareBox.State(color: .gray),
    ]
  }

  enum Action {
    case squareBox(IdentifiedActionOf<SquareBox>)
  }

  var body: some ReducerOf<Self> {
    Reduce { state, action in
      switch action {
      case let .squareBox(.element(id, .tapGesture)):
        state.squareBoxes.remove(id: id)
        return .none
      }
    }
    .forEach(\.squareBoxes, action: \.squareBox) {
      SquareBox()
    }
  }
}

struct GalleryView: View {
  var store: StoreOf<Gallery>

  var body: some View {
    ScrollView {
      LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible()), GridItem(.flexible())]) {
        ForEach(
          store.scope(state: \.squareBoxes, action: \.squareBox),
          id: \.state.id
        ) { squareBoxStore in
          SquareBoxView(store: squareBoxStore)
        }
      }
    }
  }
}

// MARK: - The App

@main
struct TheApp: App {
  var body: some Scene {
    WindowGroup {
      GalleryView(
        store: Store(initialState: Gallery.State()) {
          Gallery()
        }
      )
    }
  }
}

```

**Comment by mbrandonw** [comment, marked as answer]
> This issue was referenced in the comments for discussion https://github.com/pointfreeco/swift-composable-architecture/discussions/2694 on this forum, but without a solution. (Sorry if I've missed other references to the problem.)

Ah yes, sadly that conversation went over to email and I forgot to update the thread.

We should document this somewhere, but the problem is with any kind of lazy view, such as the `LazyVGrid` you have in your demo. Such views apparently construct view hierarchy on non-main threads, which cause stores to be created on non-main threads (due the laziness of `_StoreCollection`).

Someday in the future (once the dust settles on Swift concurrency), stores will be thread safe and this will be fine. But until then you need to eager turn your `store.scope(…)` into an array:

```swift
Array(store.scope(state: \.squareBoxes, action: \.squareBox))
```

That should fix it.

**Reply by gerfarfel** [reply]
Hi @mbrandonw, thanks for the quick response. Unfortunately, the one-line change you suggested isn't working for me – I'm still seeing the runtime warning in the modified sample code.

**Reply by mbrandonw** [reply]
The warning is different now: "Store state was accessed on a non-main thread. …". It's happening because of this: `id: \.state.id`. The `LazyVGrid` will access that key path on the background thread, hence accessing state in the store on a background thread.

If you drop the `id` argument it will stop warning. Is there a reason you wanted to provide an explicit ID? Stores already have a stable ID and so it can be used implicitly.

**Reply by gerfarfel** [reply]
Okay, thanks, that does resolve the issue in the example code. Although there's no reason to include the `id` in the example code, I did have a reason to include it in my real-world code. Now I know what the constraints are, I'll see if I can work around the problem.

**Reply by topkim993** [reply]
Hi! @gerfarfel

I'm facing the same issue. Have you found a solution? 

If so, could you please share it with me?

**Reply by gerfarfel** [reply]
Hi @topkim993 – I haven't been able to fully reconstruct my original problem because I never committed the broken code, but as best I can remember, I found a way to make the objects that the `ForEach` iterated over conform to `Identifiable`. That allowed me to drop the `id: \.state.id` from the `ForEach` and resolved the problem.

**Reply by topkim993** [reply]
@gerfarfel Thank you for your reply 😀

**Comment by larryonoff** [comment]
We frequently encounter the same crash in Firebase during release, not during testing. This crash occurred recently with the latest app release and the latest version of TCA. How can we optimize the code to resolve this issue?

<img width="1037" alt="Screenshot 2025-06-25 at 11 18 55" src="https://github.com/user-attachments/assets/c691e458-821f-44c5-937e-c219d8122245" />


```swift
    WithPerceptionTracking {
      ScrollViewReader { scroll in
        ScrollView(.vertical) {
          WithPerceptionTracking {
            LazyVStack(spacing: 0, pinnedViews: [.sectionHeaders]) {
              Section {
                ForEach(
                  Array(store.scope(state: \.templates, action: \.templates))
                ) { store in
                  WithPerceptionTracking {
                    let item = store.withState { $0 } // 🐛 the crash happens on this line
                    let isSelected =
                      self.store.selectedTemplateID == item.id

                    TemplateItemView(
                      store: store,
                      isPremium: self.store.purchases.isPremium,
                      isSelected: isSelected,
                      selectedAssetsCount: self.store.clips.count
                    )
                    .id(item.id)
                  }
                }
              } header: {
                 SomeHeader()
              }
              .onChange(of: store.selectedTemplateID) { [
                oldValue = store.selectedTemplateID
              ] templateID in
                guard let templateID else { return }

                withAnimation(oldValue == nil ? nil : .default.speed(3)) {
                  scroll.scrollTo(templateID, anchor: .center)
                }
              }
              .onChange(of: store.templates.ids) { [oldValue = store.selectedTemplateID] _ in
                guard let templateID = store.selectedTemplateID else {
                  return
                }

                withAnimation(oldValue == nil ? nil : .default.speed(3)) {
                  scroll.scrollTo(templateID, anchor: .center)
                }
              }

```

**Reply by topkim993** [reply]
Hi @larryonoff !

Let me share my experience:

### **View parameter design improvement**
Instead of passing individual values like this:
```
TemplateItemView(
  store: store,
  isPremium: self.store.purchases.isPremium,
  isSelected: isSelected,
  selectedAssetsCount: self.store.clips.count
)
```
It’s often cleaner and more reliable to include these values in the child State itself. That way, you can simply pass the scoped store and avoid pulling values from the parent store within the view.

For example:

```
// TemplateItem
struct State: Identifiable {
  var id: UUID
  var isPremium: Bool
  var isSelected: Bool
  var selectedAssetsCount: Int
  // ...
}
```

This allows your view to access everything it needs directly from store.state, reducing dependencies and making things safer.

### **Avoid using withState for view logic**
Using store.withState { $0 } like this:

```
let item = store.withState { $0 }
let isSelected = self.store.selectedTemplateID == item.id
```

can be fragile, especially when it’s used frequently during view rendering. It’s generally more robust to compute values like isSelected at the state level and pass them down through the scoped store.

### **Use .id(store.state.id) inside the view**
Applying .id(item.id) outside the view, especially when using ScrollViewReader, can sometimes lead to runtime warnings (not crashes) — particularly during initial rendering or when the ID isn't immediately available.

A safer pattern I’ve found is to assign .id(store.state.id) inside the view itself:

```
// Inside TemplateItemView body
.id(store.state.id)
```
This has helped avoid those runtime warnings in practice.

### **Be careful not to confuse store.id with store.state.id**
store.id refers to the internal ObjectIdentifier of the store instance.

store.state.id is the actual Identifiable value you define in your state (usually a UUID).

They serve different purposes, so it’s important not to mix them up, especially when working with things like ForEach(Array(store.scope(...))).

**Reply by larryonoff** [reply]
@topkim993 Thanks for sharing your experience! I expected this, but hoped there’d be alternative solutions.
